sql/schemachanger: copy current deprules to release_24_2

spilchen opened the pull request:
With the server version changing soon, we need to copy the current dependency rules for the declarative schema changer to a separate package so they remain static. New rules added to the declarative schema changer will continue to modify the current rules, which now apply to version 24.3.

The release_24_2 rules aren’t in use yet because the 24.3 server version hasn’t been implemented in the code. When that happens, it should automatically use new 'current' rules for 24.3. As well, another change is needed in current/helpers.go to update the rulesVersionKey to 24.3.

This also removes the old 23.2 rules since the minimum supported version is now 24.1.

I’m including this as part of the ALTER COLUMN TYPE epic because I need to introduce new dependency rules.

Epic: CRDB-25314
Informs: #127014
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128581)
<!-- Reviewable:end -->

spilchen committed:
sql/schemachanger: copy current deprules to release_24_2

With the server version changing soon, we need to copy the current
dependency rules for the declarative schema changer to a separate
package so they remain static. New rules added to the declarative schema
changer will continue to modify the current rules, which now apply to
version 24.3.

The release_24_2 rules aren’t in use yet because the 24.3 server version
hasn’t been implemented in the code. When that happens, another change
is needed in current/helpers.go to update the rulesVersionKey to 24.3.

This also removes the old 23.2 rules since the minimum supported version
is now 24.1.

I’m including this as part of the ALTER COLUMN TYPE epic because I need
to introduce new dependency rules.

Epic: CRDB-25314
Informs: #127014
Release note: None

fqazi reviewed (approved):
These look good <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 28 of 28 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128581#-:-O3n6Ee3BT8mpRM7qatV:b-nzma0e)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
tftr @fqazi 

bors r=fqazi

RaduBerinde commented:
Nice, thanks!

> The release_24_2 rules aren’t in use yet because the 24.3 server version hasn’t been implemented in the code. When that happens, it should automatically use new 'current' rules for 24.3. As well, another change is needed in current/helpers.go to update the rulesVersionKey to 24.3.

Yes, I am working on the change that adds `V24_3` which also does this.
